Create ColumnInfo class for mixin columns

taldcroft opened the pull request:
This is an initial implementation for #3518 to create `ColumnInfo` (aka `ColumnAttribute`) class for mixin columns.

One point is that for classes with built-in mixin support the `info` attribute is a class property wrapping `_info` as a `ColumnInfo` object, while for external classes (e.g. pandas Series), `info` is just a plain attribute which is a `ColumnInfo` object.  

Currently all the calls to `col_setattr` and `col_getattr` have been left in place, but those methods are now trivial calls to inspect the `col.info` object.  Will get around to doing the replacement once @mhvk is happy with the direction in this PR.

Closes #3518.

mhvk commented:
@taldcroft - just so you know: this is definitely on my radar and now that teaching/exams are over, I hope to get to it soonish.

taldcroft commented:
@mhvk - copy.  I took off the WIP in the title since I have now entirely removed `col_getattr` and `col_setattr`.  This is definitely a good thing in terms of the way code now looks.  The trade is that any Table column *must* have the `info` attribute/property.  I think this is worth it.

embray commented:
This looks super useful :+1:

embray commented on astropy/coordinates/sky_coordinate.py:
```diff
@@ -172,6 +173,15 @@ def __init__(self, *args, **kwargs):
+        """
+        if not hasattr(self, '_info'):
+            self._info = ColumnInfo()
+        return self._info
```
Could just use
```python
@lazyproperty
def info(self):
    return ColumnInfo()
```
:)

mhvk replied:
@taldcroft - a first comment only yet, but I thought that rather than duplicate code over all classes, one could do something like:
```
info = property(get_column_info())
```
where `get_column_info` is from `utils` as well and takes a class as an argument, i.e., something like:
```
def get_column_info(cls=ColumnInfo):
    def column_info(self):
        if not hasattr(self, '_info'):
            self._info = cls(self)
         return self._info
    return column_info
```
For `Quantity`, one would then have
```
    info = property(get_column_info(QuantityColumnInfo))
```

But I need to think through this a bit more. In particular, I worry `info` is a sufficiently general name that for some out-of-astropy classes it is going to be used already.

embray replied:
@mhvk That's what `@lazyproperty` is.  No need for that extra complexity.  If you wanted to be able to have different `ColumnInfo` subclasses that could probably live as an attribute on the `Column` class or something.

mhvk replied:
OK, I need to think better. Though we do have a different subclass in `Quantity`, but that could just be defined like you did above.

taldcroft replied:
@embray - yup, the `@lazyproperty` looks nice here.

@mhvk - I've also been a little worried about `info` as being too generic.  `attrs` is probably less common (and also less informative in code).  Anyway `col_info` would be OK by me, though you end up with a lot of `col.col_info`.  Other suggestions?

taldcroft replied:
If I understand in quantity we would just have
```
@lazyproperty
def col_info(self):
    return QuantityColumnInfo()
```

embray replied:
That would do it.  If you really wanted to go wild, if `QuantityColumn` is a subclass of `Column` you could do something in the base class like like
```python
ColumnInfoClass = ColumnInfo

@lazyproperty
def info(self):
    return self.ColumnInfoClass()
```

while `QuantityColumn` could set `ColumnInfoClass = QuantityColumnInfo` if need be.

I also agree "info" is vague, but I don't think "col_info" is particularly elucidating since if you're using that attribute you would generally already know that the object you're looking it up on is a column.  I think this is a job for documentation just specifying what goes into the ColumnInfo object.  This could be documented both in the docs for the .info property, and for the ColumnInfo object itself.

cdeil commented:
:+1: , thanks!

@taldcroft – see https://travis-ci.org/astropy/astropy/jobs/63857641#L1971

cdeil commented on astropy/table/info.py:
```diff
@@ -0,0 +1,85 @@
+    to include.  This can be a string, a function, or a list of
+    strings or functions.  Built-in options are:
+
+    - ``meta``: basic column meta data like ``dtype`` or ``format``
```
I think this docstring needs to be updated:
```
ValueError: option=meta is not an allowed information type
```

mhvk replied:
Should change to ``attributes``?

cdeil commented on astropy/table/info.py:
```diff
@@ -0,0 +1,85 @@
+    strings or functions.  Built-in options are:
+
+    - ``meta``: basic column meta data like ``dtype`` or ``format``
+    - ``stats``: basic statistics: minimum, mean, and maximum
```
Here's an example:
```
t = Table.read('http://fermi.gsfc.nasa.gov/ssc/data/access/lat/4yr_catalog/gll_psc_v16.fit')
t.info(option='stats')
/opt/local/Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/site-packages/numpy/core/_methods.py:102: RuntimeWarning: invalid value encountered in subtract
  x = asanyarray(arr - arrmean)
<Table length=3034>
        name             min          mean          std          max    
-------------------- ----------- ------------- ------------- -----------
         Source_Name          --            --            --          --
             RAJ2000      0.0377       185.638       100.585     359.881
             DEJ2000    -87.6185      -2.23514       41.5371     88.7739
                GLON    0.126161       186.273       107.572     359.969
                GLAT    -87.9516       1.22728       36.4998     86.3679
   Conf_68_SemiMajor         nan           nan           nan         nan
   Conf_68_SemiMinor         nan           nan           nan         nan
      Conf_68_PosAng         nan           nan           nan         nan
   Conf_95_SemiMajor         nan           nan           nan         nan
   Conf_95_SemiMinor         nan           nan           nan         nan
      Conf_95_PosAng         nan           nan           nan         nan
             ROI_num           1 424.777191826 238.846231996         840
          Signif_Avg        -inf          -inf           nan     1048.96
        Pivot_Energy     100.798       2104.46        2578.2     37202.0
        Flux_Density 4.21515e-16   1.29389e-11   2.17231e-10 1.14476e-08
    Unc_Flux_Density        -inf          -inf           nan 6.86476e-10
```
Stilts gives this:
```
$ stilts tpipe fermi.fits omode=stats
Total Rows: 3034
+----------------------+----------------+---------------+---------------+---------------+------+
| column               | mean           | stdDev        | min           | max           | good |
+----------------------+----------------+---------------+---------------+---------------+------+
| Source_Name          |                |               |               |               | 3034 |
| RAJ2000              | 185.63786      | 100.58523     | 0.0377        | 359.8812      | 3034 |
| DEJ2000              | -2.235143      | 41.53712      | -87.6185      | 88.7739       | 3034 |
| GLON                 | 186.27281      | 107.57225     | 0.12616104    | 359.9692      | 3034 |
| GLAT                 | 1.2272775      | 36.499836     | -87.95155     | 86.36791      | 3034 |
| Conf_68_SemiMajor    | -Infinity      |               | -Infinity     | 0.6587264     | 3031 |
| Conf_68_SemiMinor    | -Infinity      |               | -Infinity     | 0.32094488    | 3031 |
| Conf_68_PosAng       | -0.21967068    | 51.926346     | -89.94        | 89.92         | 3006 |
| Conf_95_SemiMajor    | -Infinity      |               | -Infinity     | 1.0681        | 3031 |
| Conf_95_SemiMinor    | -Infinity      |               | -Infinity     | 0.5204        | 3031 |
| Conf_95_PosAng       | -0.21967068    | 51.926346     | -89.94        | 89.92         | 3006 |
| ROI_num              | 424.7772       | 238.84624     | 1             | 840           | 3034 |
| Signif_Avg           | -Infinity      |               | -Infinity     | 1048.9587     | 3034 |
| Pivot_Energy         | 2104.4626      | 2578.198      | 100.79754     | 37201.97      | 3034 |
| Flux_Density         | 1.29388835E-11 | 2.1723061E-10 | 4.2151523E-16 | 1.1447582E-8  | 3034 |
| Unc_Flux_Density     | -Infinity      |               | -Infinity     | 6.864756E-10  | 3034 |
```
I think putting min / max next to each other is nicer.

Handing nan / inf in the stats computations is tricky , but generally I think having a column like "good" that prints what `np.isfinite` gives would be very helpful.

taldcroft commented:
@cdeil - thanks for the useful testing and feedback.

The latest commits do:
- Add a `shape` column to the `attributes` option if there is any column with non-trivial shape.  This shows up in your test file.
- Add a `n_bad` column at the end if there are any columns with at least one "bad" element.  For a masked table "bad" means masked.  For a non-masked table "bad" means any `nan` or `inf` elements.  If you have different ideas for defining the number of bad I'm open to suggestion.
- I chose "bad" instead of "good" because I personally think that is easier to interpret since bad values are more often in the minority (zero or a few).
- Note: if there are no bad elements in the table then the `bad` column (with all zeros) is not included.  I was wondering if this might be confusing, but it follows the idea of this tool that outputs are suppressed if they are all trivial.
- Reordered stats per your suggestion.

I'm going to hold off on updating tests and docs until there is agreement that the outputs are OK.

cdeil commented:
+1 to `n_bad` instead of `n_good` and to show the shape.

Can you change the default stats functions so that one gets e.g. a min / max for `Conf_68_PosAng` in the Fermi catalog example table? Currently I always get `nan` if there's a `nan` in a float column:
```
 Conf_68_SemiMajor           nan           nan         nan         nan    28
   Conf_68_SemiMinor           nan           nan         nan         nan    28
      Conf_68_PosAng           nan           nan         nan         nan    28
   Conf_95_SemiMajor           nan           nan         nan         nan    28
   Conf_95_SemiMinor           nan           nan         nan         nan    28
      Conf_95_PosAng           nan           nan         nan         nan    28
  ```

taldcroft commented:
@mhvk - the last commit is a possible implementation for formatting mixins properly within tables:
```
In [2]: t = QTable([[1,2]*u.m, SkyCoord([1,2], [1,2], unit='deg,deg')])

In [3]: t
Out[3]: 
<QTable masked=False length=2>
  col0    col1 
   m    deg,deg
float64  object
------- -------
    1.0 1.0 1.0
    2.0 2.0 2.0
```

taldcroft commented:
@cdeil - I think 49af58f does the trick.  This gives a `stats` output that looks similar to stilts, except for having `nan` in some places where stilts has a blank.

cdeil commented on astropy/table/info.py:
```diff
@@ -0,0 +1,88 @@
+        Info option (default='attributes')
+    out: file-like object, None
+        Output destination (default=sys.stdout).  If None then the
+        OrderedDict with information attributes is returned
```
@taldcroft – I think the `None` option to return the `OrderedDict` isn't implemented yet, right?
```
In [23]: t.info(out=None)
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-23-ec9644a5a55b> in <module>()
----> 1 t.info(out=None)

/Users/deil/code/astropy/astropy/table/info.py in info(self, option, out)
     86 
     87     outlines.extend(info.pformat(max_width=-1, max_lines=-1, show_unit=False))
---> 88     out.writelines(outline + os.linesep for outline in outlines)

AttributeError: 'NoneType' object has no attribute 'writelines'
```
I think it would be useful to be able to get a Table or OrderedDict back, so that it's possible to e.g. store the stats.

mhvk replied:
A table would seem best, since this is created anyway!

taldcroft replied:
Done in 0455842.

mhvk commented:
@taldcroft - one nice thing about meetings is that sometimes one gets to skip a session and do some astropy...  I now had a fairly detailed look and I like this `ColumnInfo` PR very much. Indeed, so nice that I have no bigger comments than that I wonder why `column_info.py` is in `astropy.utils` rather than `astropy.table`. (More detailed comments will follow in-line.)

p.s. With the combination of getting information from columns, `info` now is, I think, just the right name.

mhvk commented on astropy/units/quantity.py:
```diff
@@ -114,6 +115,16 @@ def __next__(self):
+class QuantityColumnInfo(ColumnInfo):
+    _attrs_from_parent = set(['dtype', 'unit'])
+
+    def __init__(self, *args, **kwargs):
```
This call should not be necessary, right?

taldcroft replied:
Ah yes, it's a leftover from when `__init__` here was doing something.  I just blindly deleted that one line.

mhvk commented on astropy/units/quantity.py:
```diff
@@ -114,6 +115,16 @@ def __next__(self):
+    def __init__(self, *args, **kwargs):
+        super(QuantityColumnInfo, self).__init__(*args, **kwargs)
+
+    def default_format_func(self, format, val):
```
I wonder whether it wouldn't be clearer conceptually to just call this `default_format` and let it have exactly the same meaning as `Column.format`, i.e., if it is callable, it is a function, and if not, it is tried to interpret it as a format string.  I realise this makes it slightly slower to print, but I think it is worth the clarity.  With that and the comment above, one would just have
```
class QuantityColumnInfo(ColumnInfo):
    _attrs_from_parent = set(['dtype', 'unit'])
    default_format = '{0.value}'
```

taldcroft commented:
>  I wonder why column_info.py is in astropy.utils rather than astropy.table

That allows `Quantity`, `Time` et al. to import `ColumnInfo` without importing all of `astropy.table`.  I struggled with this a bit but as far as I know you can't get `astropy.table.whatever` without `astropy.table.__init__` running.

mhvk commented on astropy/coordinates/sky_coordinate.py:
```diff
@@ -41,6 +42,24 @@ def FRAME_ATTR_NAMES_SET():
+            data = self._parent_col().data
+            self.unit = ','.join(str(getattr(data, comp).unit) for comp in data.components)
+
+    def default_format_func(self, format, val):
```
Here, clearly a function is useful, but I'm a bit confused about the need for a cache. Is this for the case that one changes the representation half-way through, and then perhaps change it back? Since this quite a list comprehension loop, is this worth it? Could one just calculate it on the fly?

Separately, I think it would in principle be nice if `col.info.default_format` gave a string that would give a good clue on how to set `format` to do something else. Could one do this using my `default_format` suggestion and implement it as a property as follows?
```
@property
def default_format(self):
    if self.parent_col is None:
       raise ValueError("Default format is undefined for a SkyCoordColumnInfo "
                        "instance not attached to a SkyCoordObject.")
    formats = ['{0.' + compname + '.value:}' for compname
               in self._parent_col().data.components]
    return ' '.join(formats)
```

mhvk replied:
p.s. Is `.value` is necessarily the best option? Since the components could be `Angle`, one could also get hhmmss etc...  But that probably is best left for later.

mhvk commented on astropy/units/quantity.py:
```diff
@@ -607,6 +618,15 @@ def to(self, unit, equivalencies=[]):
         return self._new_view(new_val, unit)
 
     @property
+    def info(self):
```
@embray earlier suggested to use `lazyproperty` here, so that this is shorter. I think that makes sense:
```
@lazyproperty
def info(self):
    return QuantityColumnInfo(self)
```

taldcroft replied:
Yes, I actually did implement `@lazyproperty` across the board, but then realized that `lazyproperty` actually just sets the value in the object dictionary.  That means that `info` is not (AFAIK) an actual property and it's not read-only (so there is a chance to mess it up saying `col.info = 5`).

However, there was a more important and subtle problem which actually required going away from `lazyproperty` and now I don't remember the details.  Maybe it was related to copying columns on initialization.

mhvk replied:
OK, fair enough. Eventually we may want to simplify this a bit to avoid needless boilerplate, but no need to do that now.

mhvk commented on astropy/table/column.py:
```diff
@@ -110,11 +55,15 @@ def _col_update_attrs_from(newcol, col, exclude_attrs=['name', 'parent_table']):
 
-    attrs = COLUMN_ATTRS - set(exclude_attrs)
+    if not hasattr(newcol, 'info'):
+        info_attr = '_info' if hasattr(col.__class__, 'info') else 'info'
```
I'm a little confused why we check both the object and class for `info` here, and why it would help to write to `_info` is the object does not have `info` but the class does.

mhvk commented on astropy/table/column.py:
```diff
@@ -131,25 +80,28 @@ def col_iter_str_vals(col):
-    if hasattr(col, '_astropy_column_attrs'):
-        col_setattr(col, 'parent_table', None)  # Don't copy weakref to parent table
-    newcol = col.copy() if hasattr(col, 'copy') else deepcopy(col)
+    col.info.parent_table = None  # Don't copy weakref to parent table
```
This seems risky: aren't you somehow detaching the column to be copied from its parent table here? Shouldn't one just avoid copying that particular link?

taldcroft replied:
Good catch.  Don't know what I was thinking, except that I was planning to restore and forgot.

If I leave the `parent_table` weakref there then it generates an infinite recursion.  I haven't drilled down into this, but the new code in 4356f78 temporarily removes the parent_table link and puts it back in a safe way.

mhvk commented on astropy/table/column.py:
```diff
@@ -131,25 +80,28 @@ def col_iter_str_vals(col):
     return newcol
 
 
+def add_column_info(col):
```
I don't think this is actually used anywhere, or am I mistaken?

taldcroft replied:
It's used in `astropy/table/tests/conftest.py`.  This is intended to be the public method for making some array-like object a mixin column.

mhvk replied:
Since it is not yet directly needed, my tendency would be to omit this for now, and simply explain what one would have to do in the documentation (maybe with a warning that this is still subject to change). 

If we do keep it, we should raise an exception if `info` is already present.

mhvk commented on astropy/table/pprint.py:
```diff
@@ -384,8 +383,11 @@ def _pformat_col_iter(self, col, max_lines, show_name, show_unit, outs,
-        col_format = col_getattr(col, 'format')
-        format_func = _format_funcs.get(col_format, _auto_format_func)
+        col_format = col.info.format
+        if col_format is None and hasattr(col.info, 'default_format_func'):
```
With my comments on changing to `default_format`, this would become
```
col_format = col.info.format or getattr(col.info, 'default_format', None)
format_func = _format_funcs.get(col_format, _auto_format_func)
```

taldcroft replied:
@mhvk - I had not read your proposed code carefully before, but was thinking about this offline now and then over the last few days (busy with other stuff).  I basically wrote the same code in my head, so we are agreed!

mhvk replied:
OK, good.  Seems that nearly covers it!

mhvk commented:
> That allows `Quantity`, `Time` et al. to import `ColumnInfo` without importing all of `astropy.table`.  I struggled with this a bit but as far as I know you can't get `astropy.table.whatever` without `astropy.table.__init__` running.

OK, that makes sense. Before you had added the statistics, I might have argued to just put `QuantityColumnInfo`, etc., under `table`, but with the statistics the `info` property becomes useful also outside of a table, so in that respect putting it under `astropy.utils` is actually logical. And I see that the only actual reference to anything from `table` is to look at whether something is a `BaseColumn` instance. Maybe one can just check on class name equals `Column` or `MaskedColumn`? Anyway, that really is a detail; bottom line: location makes sense (maybe even to the extent that we should change the name...).

taldcroft commented:
@mhvk @embray - after saying there was some subtle problem in using `lazyproperty`, I tried to reproduce the problem and could not!  After fc98a54 it passes tests (except some doctests I haven't fixed up yet).  Go figure.  Anyway, I agree this simplifies things nicely so we'll run with it unless something else comes up.

taldcroft commented:
> (maybe even to the extent that we should change the name...)

That's interesting.  You have a point that this has gone somewhat beyond just columns and tables.  Maybe `DataInfo`?  I wonder if this would have any place in `NDData`.

taldcroft commented:
> Maybe one can just check on class name equals Column or MaskedColumn?

I think we want to check for subclasses.  In theory you can drill down the MRO and do string checks, but I don't think that's worth the bother.

mhvk commented:
@taldcroft - +1 to `DataInfo`. It might well be useful to `NDData` as well, though I think it is more general so fine to keep it in `utils`. (An `NDData` array could be a column, with, e.g., each plane being a time-stamped image...)

mhvk commented on astropy/utils/column_info.py:
```diff
@@ -0,0 +1,232 @@
+    from ..table.column import BaseColumn
+
+    if attr == 'class':
+        val = '' if isinstance(col, BaseColumn) else col.__class__.__name__
```
This is the only place where `BaseColumn` is used, correct? But are you sure that you would not want class names to appear for subclasses of `Column` and `MaskedColumn`? It seems to me those would be unusual and thus worth listing. If so, we can still remove all dependence on `table` here by checking `col.__class__.__name__`.

mhvk commented on astropy/utils/column_info.py:
```diff
@@ -0,0 +1,232 @@
+    return str(val)
+
+
+class ColumnInfo(object):
```
Actually, maybe even simpler, if we do go for `DataInfo` here, is to just remove the check on `class` altogether and in `column.py` construct a proper `ColumnInfo(DataInfo)` that special-cases `class`.

taldcroft commented:
Now I'm running into a problem with `@lazyproperty` that I can document.  First here is a skeleton of the `ColumnInfo` class:
```
import weakref
from astropy.utils import lazyproperty

class Info(object):
    def __init__(self, parent):
        self.parent = weakref.ref(parent)

class MyObj(object):
    @lazyproperty
    def info(self):
        return Info(self)
```
Now I create a `MyObj` object and do a deepcopy:
```
In [11]: m = MyObj()

In [12]: id(m)
Out[12]: 4357664400

In [13]: id(m.info)
Out[13]: 4357664528

In [14]: id(m.info.parent)
Out[14]: 4357696072

In [15]: id(m.info.parent())
Out[15]: 4357664400

In [16]: mc = deepcopy(m)

In [17]: id(mc)
Out[17]: 4357664656

In [18]: id(mc.info)
Out[18]: 4357663952

In [19]: id(mc.info.parent)
Out[19]: 4357696072

In [20]: id(mc.info.parent())
Out[20]: 4357664400
```
So the deep copy points to the wrong parent.  In some sense `deepcopy` has done what you would expect, but it isn't what we need.  What has me more confused is that in other cases the deep-copied weak reference is just `None`.  For instance if I deepcopy an instance of `table_helpers.ArrayWrapper`:
```
In [10]: from astropy.table.table_helpers import ArrayWrapper

In [11]: a = ArrayWrapper([1,2,3])

In [12]: a.info._parent_col
Out[12]: <weakref at 0x106fe0730; to 'ArrayWrapper' at 0x106fe2910>

In [13]: ac = deepcopy(a)

In [14]: print ac.info._parent_col
None
```
Anyway, I can solve this by going back to the original non-lazy property for `info` and then having that basically set `_parent_col` with `self` (i.e the parent column) every time it is called, thus ensuring that things line up whenever `info` is accessed.

@mhvk @embray - thoughts?

mhvk commented:
Hmm, this is odd. It seems related to what I would consider a bit of a bug in `lazyproperty`, viz., that it effectively just assigns `m.info = ColumnInfo()`. As a result, unlike what is the case for a regular propery, one can simply override it: `m.info = 'bla'`.

taldcroft commented:
My previous comment is hidden now, but this was part of my original objection to `lazyproperty`:

> Yes, I actually did implement @lazyproperty across the board, but then realized that lazyproperty actually just sets the value in the object dictionary. That means that info is not (AFAIK) an actual property and it's not read-only (so there is a chance to mess it up saying col.info = 5).

taldcroft commented:
@mhvk - what do you think about 1c6eacb which now uses a descriptor for defining the `info` property of class.  Implications:

- Cleaner syntax overall, less boilerplate
- For a class where you use `add_column_info`, this injects `info` as a class descriptor and is not localized to the instance.  This isn't my preference, but I think it is acceptable.

taldcroft commented:
This is passing tests locally now for py 2.6, 2.7, 3.4 and is possibly feature-complete now.  Some highlights of recent commits:

- `pandas.Series` has a `mask` method and this conflicts with the mixin injection of a mask `FalseArray`.  I'm not sure why this was working previously, but for now I'm thinking to not worry about pandas for mixins.  That was kinda nifty but probably fragile in reality.  Maybe it could be fixed later, but it has been dropped.
- Made the repr for `col.info` be the same as the `col.info()` output, so that's effectively a shortcut.  Doing the same for `table.info` would be nice.
- Added `length` as a standard column attribute

mhvk commented:
@taldcroft - I like the use of a `Descriptor`! But less fiddling with someone's class. Is there anything against continuing to just stick on `info` for a one-off case? I realise that means it can get overwritten, but I'd see that as the lesser evil.

Alternatively, how about just postponing `add_column_info` -- I think with this PR we get a *very* nice method for making a class compatible with `Table` (and at the same to get general information about an object; great!). Shall we worry about how one adds individual objects later?

taldcroft commented:
Agreed. Adding to a individual object was always a bit hacky and fragile.

taldcroft commented:
@mhvk - I think I've addressed all your comments, including using the more generic DataInfo / data_info everywhere and:
> Maybe one can just check on class name equals Column or MaskedColumn? 

AFAIK the only thing left are narrative docs.

mhvk commented:
@taldcroft: to save you looking through the travis logs:
* `sphinx` failed because "/home/travis/build/astropy/astropy/build/lib.linux-x86_64-2.7/astropy/table/info.py:docstring of astropy.table.Table.info:8: WARNING: Inline literal start-string without end-string."
* numpy <1.8 failed because `np.nanmean` was only added to numpy in 1.8.

For the latter, we could either use regular instead of `nan` functions, make a quick hack, or backport `nanmean` (and perhaps all `nanfunctions`) to `utils.compat.numpy`. I think I'd go for the hack myself... (I looked at backporting, but quite a few different things changed, including where those functions are located; it didn't quite seem worth it).

taldcroft commented:
@mhvk - thanks.  I have patched these locally taking the hack approach of using the regular functions for numpy < 1.8.  I also need to fix one test that fails for numpy < 1.8 since the info output is different without the nan funcs.

mhvk commented on astropy/coordinates/sky_coordinate.py:
```diff
@@ -13,6 +13,7 @@
 from .. import units as u
 from ..wcs.utils import skycoord_to_pixel, pixel_to_skycoord
 from ..utils.exceptions import AstropyDeprecationWarning
+from ..utils.data_info import DataInfo, BaseInfo
```
The names are confusing to me, as they don't make explicit the roles that they have, i.e., one has to know that `DataInfo` is a descriptor, and `BaseInfo` the base for the class passing back information. It doesn't help to call the actual class `SkyCoordDataInfo`. I think I'd go with (in order of strongly to weakly suggested):
* renaming `SkyCoordDataInfo` to `SkyCoordInfo`
* maybe renaming `DataInfo` to `InfoDescriptor` (yes, duplicate but makes point clearer)
* If so, renaming `BaseInfo` to `DataInfo`

I think the code below is clearer if one sees
```
    info = InfoDescriptor(SkyCoordInfo)
```

taldcroft replied:
At some point I had a mental note to use `<classname>Info` but forgot to implement.  I was also struggling with naming, but I think you've hit on a good scheme so I'll do that.

mhvk commented on astropy/table/column.py:
```diff
@@ -131,21 +78,19 @@ def col_iter_str_vals(col):
-        col_setattr(col, 'parent_table', None)  # Don't copy weakref to parent table
-    newcol = col.copy() if hasattr(col, 'copy') else deepcopy(col)
+    parent_table = col.info.parent_table
+    col.info.parent_table = None
```
Since it took a lot of time to figure out it was necessary, maybe add a comment here why you are temporarily removing the weakref and doing the `try/finally`?

mhvk commented on astropy/time/core.py:
```diff
@@ -112,6 +113,23 @@
 
+class TimeInfo(BaseInfo):
+    """
+    Make dtype and unit be None and be read-only.  Setting attrs_from_parent
```
You mention `dtype` here but don't define it!

mhvk commented on astropy/time/core.py:
```diff
@@ -112,6 +113,23 @@
+        return None
+
+    info_summary_stats = staticmethod(
+        data_info_factory(names=BaseInfo._stats,
```
Does this work currently? in particular, it seems one should not be able to get a `mean` yet:
```
np.mean(Time(['J2000.', 'J2001']))
# OperandTypeError: Unsupported operand type(s) for +: 'Time' and 'Time'
```

taldcroft replied:
`mean` fails, as expected (cannot add two times).

mhvk replied:
Yes, that was what I worried about. But of course, it is captured in the try/except in the function produced by `data_info_factory`... So this doesn't matter, sorry.

mhvk commented on astropy/units/quantity.py:
```diff
@@ -114,6 +115,13 @@ def __next__(self):
+class QuantityInfo(BaseInfo):
+    attrs_from_parent = set(['dtype', 'unit'])
+
+    def default_format_func(self, format, val):
```
Still `default_format_func`!

mhvk commented on astropy/utils/data_info.py:
```diff
@@ -0,0 +1,291 @@
+    from ..table.column import Column, MaskedColumn
+
+    if attr == 'class':
+        val = '' if type(col) in (Column, MaskedColumn) else type(col).__name__
```
Still checking class instead of name!   Would replace with:
```
val = '' if type(col).__name__ in ('Column', 'MaskedColumn') else type(col).__name__
```

mhvk replied:
Actually, if we do want to propagate this to `Column` instances, I guess one could remove the test altogether if in `ColumnInfo` one defined
```
_info_summary_attrs = ('dtype', 'shape', 'unit', 'format', 'description')
info_summary_attributes = staticmethod(...)
```
(Sorry if not, I'm still a little confused how this all works.)

Should add that I think I like the check by name better.

Or perhaps just always list the class? For QTable, `Column` would be the unusual one, though still useful for strings.

taldcroft replied:
I don't want to check by name because there is no reason that a user-defined mixin class cannot be named `Column` (even if it would be a perverse name choice).  The only 100% reliable check is by matching the type.

One issue is that this function is Table / Column specific and needs to be generalized or at least overridable anyway.  It probably needs to be a class method, but I need to think about it a little.

embray replied:
I haven't been following this PR very closely for a while, but is it possible [`OrderedDescriptor`](https://github.com/astropy/astropy/pull/3679) could be useful for this?  It seems like a lot of its functionality is maybe being duplicated here.

embray replied:
Actually, ignore that, I'm not entirely sure how it would be useful.  I'll study up on this PR more before making any more suggestions :)

mhvk replied:
@taldcroft - on further thought, isn't it most logical for `col.info` to actually list the class name? I can see that you might want to suppress it in `table.info` or perhaps anytime the column is part of a table, but if so, I think it makes most sense to put this directly in `table/info.py` or in `ColumnInfo`.

Anyway, obviously no big deal!

mhvk replied:
And if we suppress the class if it is all the same for a given table, I think this stanza can go, which means the import and column class check can go!

mhvk commented on astropy/coordinates/sky_coordinate.py:
```diff
@@ -41,6 +42,37 @@ def FRAME_ATTR_NAMES_SET():
+    attrs_from_parent = set(['unit'])
+
+    @staticmethod
+    def default_format(val):
```
My original idea here was to make this a property, which returns a format string that will be parsed by `pprint`, i.e., something like the following:
```
@property
def default_format(self):
    components = self._repr_data.components
    formats = ['{0.' + compname + '.value:}' for compname in components]
    return ','.join(formats)
```
I think this might not work since you seem to change the representation in `_repr_data`, but I don't understand `SkyCoord` well enough to see why this is needed; it seems you do something special mostly for sperical without a distance, but in that case I'd think one could do with just not including the distance among the components.

Anyway, mostly in case this can be done more simply;

taldcroft replied:
There is some messiness due to a design decision at the coordinate Frame level which makes the user interface more intuitive at the expense of code nastiness.  Basically you can say the Frame has a SphericalDistance representation, but if you don't provide a distance at the time of creation then the actual data representation is UnitSphericalDistance.  In this special case there are shenanigans to make the repr correct.

But the other important bit is that the Frame (or SkyCoord) representation is mutable, but the actual data values (`skycoord.frame._data`) are never touched.  So a different representation of the coord data are generated on the fly (and cached) as needed.  That's what's happening in `_repr_data`.  This code is basically taken from the Frame `__repr__`.

Your idea about making `default_format` be a property returning a string does have merit in showing the user something useful that they can build on.  There is a detail in your implementation that would need fixing, which is that the `repr_data` components are things like `lon`, `lat` while what would be needed in the format statement are the frame names like `ra`, `dec`.  There is just a translation.

The bigger issue is about performance.  As you probably know iterating over a SkyCoord one element at a time is super slow because it creates a full-blown SkyCoord each time.  I have an idea to allow a format function to take an additional argument which is the index.  If there, it was assume the prior arg is the entire column.  So in coordinates it is *much* faster to do `c.ra[i]` than `c[i].ra`.  So using a function here instead of the format string would allow this optimization.  But I want to save that for a future PR.  Maybe speed is not the most crucial thing, but it's certainly possible someone could be writing a million-row table with a coordinate in `io.ascii`.

mhvk replied:
OK, I see how this would get messy! I think it is fine to leave it for future optimization, since nothing you do here prevents us from turning this into a string or a different function later.

mhvk commented on astropy/table/column.py:
```diff
@@ -44,77 +45,23 @@ def _auto_names(n_cols):
 
-    attrs = COLUMN_ATTRS - set(exclude_attrs)
+    if not hasattr(newcol, 'info'):
+        newcol.__class__.info = DataInfo(BaseInfo)
```
Shouldn't this have become
```
newcol.info = BaseInfo()
```

taldcroft replied:
The `__init__` for `DataInfo` takes the class not an instance.

mhvk replied:
I meant that for instances, we don't really want to touch their class at all, so why not just do what `DataInfo` does (effectively), stick on an `info` attribute. Of course, this would not be a property and thus could be overwritten, but I think that disadvantage is less than sticking something on a parent class. (I just can't bear the thought of working on some new fancy class and suddenly a property appears on it that isn't even in the definition -- how does one debug that?).

mhvk replied:
p.s. It may mean it should be
```
info = BaseInfo(newcol)
```

taldcroft replied:
Actually now I realize that since we are removing `add_column_info` and effectively the capability to have an instance-only `info` attribute, then these statements can just be deleted.

In other words, a mixin column is now required to have an `info` property as part of the class definition, and we won't monkey-patch it in.  Maybe in the future we'll do something like that, but it won't be supported functionality in this PR.

mhvk commented on astropy/table/column.py:
```diff
@@ -122,7 +69,7 @@ def col_iter_str_vals(col):
 
     Warning: this function is subject to change or removal.
```
You're removing the warning below. Is it still appropriate here?

mhvk commented:
@taldcroft - I'll try to do a final review soon.  For now, I quickly checked the failures: these are for numpy <=1.7, where tests fail on the items where `nan` is included. (Shouldn't these work on numpy = 1.7?)

taldcroft commented:
Gack, thought that was fixed.

mhvk commented on astropy/table/info.py:
```diff
@@ -0,0 +1,97 @@
+"""
+Table method and functions related to providing information about
+columns and tables.
```
I think this module is now only for tables, correct?  If so, correct docstring here as well as the docstring of `info` below.

mhvk commented on astropy/table/info.py:
```diff
@@ -0,0 +1,97 @@
+    if 'length' in info.colnames and np.all(info['length'] == len(self)):
+        del info['length']
+
+    if out is None:
```
I would move this before the removals of `n_bad` and `length`, since I would guess the option to return the table would mostly be used if one wants to look at `info` programmatically, in which case it is easier if one always gets the same table columns (even if not all of them are very useful). But for the output it is very useful to remove them.

mhvk commented on astropy/utils/data_info.py:
```diff
@@ -0,0 +1,296 @@
+import os
```
Maybe note somewhere in this file that it is extensively tested as part of `table.tests`, not in `utils.tests`?

mhvk commented:
@taldcroft - had another fairly thorough review.  Apart from minor comments, mostly about docstrings, it all looks great!

mhvk commented:
@taldcroft - I think eventually we'd want a bit more documentation on `info` outside of the table context. My feeling, though, would be to first let this shake out for one release and use it formally only for the dual purpose of making mixin possible and provide table info. This way changes are still relatively easy if needed. But if so, we should have a new issue (1.2 milestone) for adding documentation to `Quantity`, `Time`, etc.  

An example of something neat that may now become easy would be to have a `plot(t['x'], t['y'])` that automatically uses the `name` and `unit` attributes for the axes!

taldcroft commented:
Fingers crossed for tests passing, I think all comments have been addressed.
The only remaining cleanup is to remove commit a5123bb and rebase on #3824 when that is merged.

@mhvk - thanks for all your input, this has been a huge help!

I also had an idea that `Table.info` could be made an `InfoDescriptor` object where the current `table.info.info()` function is the callable method of a class like for columns.  That way you can type `tbl.info` as a shortcut for `tbl.info()`.  This also provides the opportunity to add a `_repr_html_` method for ipython notebook.  I can open an issue if you think these are good ideas.

mhvk commented:
@taldcroft - OK, great!  Let me know when the rebasing/removal is done and I'll have a last look before pressing the merge button...

cdeil commented:
Would `table.info` output here help understanding users how `Time`, `TimeDelta` and `Quantity` are stored? (see #3832)

I don't want to create extra work here, this is awesome and I want to start using it. But if it's useful, an "advanced example" for the new functionality added here could be added to the docs.

taldcroft commented:
@cdeil - I have a follow-on PR #3825 for additional docs related to mixins.  I'll keep your comments from #3832 in mind.

taldcroft commented:
@cdeil - with `table.info` you would start noticing which columns are `Column` and which are mixins because the class is explicitly listed for mixins.  Once this PR gets merged then some real-world testing might reveal slight improvements in presentation.  Right now the class for `Column`-based objects is suppressed to reduce clutter and redundant info, but it isn't totally clear if that is perfect.

mhvk commented:
@taldcroft - your note made me wonder: what about suppressing the class information only if all the columns are of the same type?

taldcroft commented:
@mhvk - that might make sense.

taldcroft commented:
From #3832:

> And I like the idea of somehow using the scale and format instead of the unit for Time and TimeDelta. With the new info property, those kinds of things become quite possible (but may need more thought).

@mhvk - I had implemented having `format` be in the `attrs_from_parent` set and thought this was all quite clever since it does show up as you would want in the info and repr.  Alas this introduced a problem because `col.info.format` always takes precedence over `default_format`:
```
col_format = col.info.format or getattr(col.info, 'default_format', None)
```
With this current logic it was using something like `iso` as a format() string and crashing.  So we could change the precedence, but that means it is not possible to override `default_format` for mixin columns that define it.  That doesn't seem great, so we need something else.  Ideas?

mhvk commented:
@taldcroft - hmm, I think there are two possible solutions: the simplest would be (for now) not to include it in `attrs_from_parent`. The slightly less simple (but not too difficult one), would be to give time a `__format__` attribute.  Something like
```
def __format__(self, format_spec):
    if format_spec in self.FORMATS:
        return str(getattr(self, format_spec))
    else:
        return format(str(self), format_spec)
```


I made a very quick check with exactly that and it seemed to work:
```
In [1]: from astropy.time import Time

In [2]: tn = Time.now()

In [3]: '{0}'.format(tn)
Out[3]: '2015-06-10 20:57:06.278625'

In [4]: '{0:fits}'.format(tn)
Out[4]: '2015-06-10T20:57:06.279(UTC)'

In [5]: td = Time.now() - tn

In [6]: '{0}'.format(td)
Out[6]: '9.064564814809106e-05'

In [7]: '{0:sec}'.format(td)
Out[7]: '7.831783999995068'
```

This means, though, one cannot freely assign to `info.format`. Also, in some sense it is overkill, since one can already do `'{0.fits}'.format(tn)`, which allows more flexibility (as in: `'{0.mjd:9.3f}'.format(tn)`.

I guess my suggestion for now would be to go the simple route and not get `format` from parent; in the end, `Time.format` is a representation, not a proper format.

taldcroft commented:
Brilliant!  Can we have our cake and eat it too with a slight mod?
```
def __format__(self, format_spec):
    if format_spec in self.FORMATS:
        return str(getattr(self, format_spec))
    else:
        return format(self, format_spec)  # <==
```
This would require some slight trickery in the `TimeInfo` class to make `info.format` default to the parent attribute but be assignable.  Then all options are open, but maybe I didn't think all the way through this.

mhvk commented:
@taldcroft -- that causes infinite recursion...

I think logically one should do:
```
def __format__(self, format_spec):
    if format_spec in self.FORMATS:
        return str(getattr(self, format_spec))
    else:
        return super(Time, self).__format__(format_spec)
```
Although this gives a `TypeError` for non-empty formats.  But that is just as it is now: `'{0.mjd:8.3f}'.format(tn)` works only because the attribute is taken before trying to find a `__format__` method.

mhvk commented:
@taldcroft - how about punting on `TimeInfo.format` for now (i.e., removing from `attrs_from_parent`)? This PR is very large as it is, and we can return to this once it is in -- I'm not sure I like my suggestion all that much myself!

mhvk commented:
p.s.  Should be specific -- since all tests pass, if there is no real issue with `format` being  in `attrs_from_parent`, them I'm also happy with just sticking with what we have.

taldcroft commented:
Agreed, this has been punted off to #3842.

mhvk commented:
@taldcroft - just checking: you were still waiting to rebase on #3824, correct?

taldcroft commented:
@mhvk - Interesting, it does not show a merge conflict even though #3824 is merge (and thanks BTW).  Maybe that's because it is the exact same commit in both?  I don't see any real reason to rebase just to get rid of that one duplicate commit if it is merging OK, do you?

mhvk commented:
@taldcroft - I just checked on my own machine and it does seem one can just merge the two. I'll do a last look-over before pressing the big green button...

mhvk commented on astropy/table/info.py:
```diff
@@ -0,0 +1,99 @@
+    to include.  This can be a string, a function, or a list of
+    strings or functions.  Built-in options are:
+
+    - ``meta``: basic column meta data like ``dtype`` or ``format``
```
Oops -- I think this should be "attributes", no?

mhvk commented on astropy/table/info.py:
```diff
@@ -0,0 +1,99 @@
+    if 'length' in info.colnames and np.all(info['length'] == len(self)):
+        del info['length']
+
+    outlines.extend(info.pformat(max_width=-1, max_lines=-1, show_unit=False))
```
If you are changing the docstring anyway, do you want to go ahead and remove the `class` column if it is all the same class?

taldcroft replied:
The "all-identical" criterion seems a bit risky.  What if you have a 3-column table with all `Time` objects?

So what about dropping only if all columns are `Column` or `MaskedColumn`?  This is nice because it takes the special-case out of `data_info.py` and puts it into `table/info.py` where it belongs.  We would change from:
```
def _get_column_attribute(col, attr=None):
    """
    Get a column attribute for the ``attributes`` info summary method
    """
    from ..table.column import Column, MaskedColumn

    if attr == 'class':
        val = '' if type(col) in (Column, MaskedColumn) else type(col).__name__
    elif attr == 'dtype':
        val = col.info.dtype.name
...
```
to
```
def _get_data_attribute(col, attr=None):
    """
    Get a data attribute for the ``attributes`` info summary method
    """
    if attr == 'dtype':
        val = col.info.dtype.name
...
 ```
(And now that I look at this there should probably be a global replace of `col` with `data` in `data_info.py`).

mhvk replied:
Yes, I like that change to `_get_data_attribute`.

As for when to omit the class: maybe it should be only if all columns are of the same type and are not mixin? 
```
if 'class' in info.colnames:
    cols = self.columns.values()  # maybe define above, before earlier loop over columns
    if not (self._is_mixin_column(cols[0]) and
            all(type(col) is type(cols[0]) for col in cols[1:]):
        del info['class']
```

mhvk commented:
OK, three minor comments (though the second two may affect a test or two...).

taldcroft commented:
Side note: I see in email that you said basically the same thing as I did (yay!), but it's not showing up in my view of this PR.  I was going to ask about the "three" comments when I only see two.  Did you delete it?

taldcroft commented:
@mhvk - OK, now I see your 3rd comment, but it's buried at the 1/2 way point of this huge conversation.

mhvk commented on astropy/table/info.py:
```diff
@@ -0,0 +1,99 @@
+        if np.all(info[name] == ''):
+            del info[name]
+
+    if np.all(info['n_bad'] == 0):
```
Should this be:
```
if 'n_bad' in info.colnames and np.all(info['n_bad'] == 0):
```

taldcroft replied:
Yes.

taldcroft replied:
And speaking of which there needs to be a handler for the empty Table case.

taldcroft commented:
@mhvk - I need to add a test of the code that removes the `class` column, but ran out of time before some meetings today.  Will add the test this evening.  In the meantime you can look at a900297.

taldcroft commented:
Ack, rushing and made one mistake that is fixed in bbb76e4.

mhvk commented:
OK, looks good so far!

taldcroft commented:
@mhvk - added 4 more commits, including 41416c4 that makes `Table.info` into an `InfoDescriptor` to give more symmetry with column info.  Fingers crossed on tests, I ran a number of configurations locally.

mhvk commented:
@taldcroft - travis fails since `Column().info` now gives the class among its properties (correctly, I think, so the test needs adjusting).

But I was surprised by your expectation that a `QTable` of two quantity columns should list the class. I think this is actually a bug in `_is_mixin_column` right now:
```
In [24]: Table()._is_mixin_column([10.]*u.m)
Out[24]: False

In [25]: QTable()._is_mixin_column([10.]*u.m)
Out[25]: True
```
This should be the other way around, right? If so, just swap `True` to `False` in the two definitions.

While you are at it: maybe also redefine `has_mixin_columns` to:
```
    @property
    def has_mixin_columns(self):
        """
        True if table has any mixin columns (defined as columns that are not Column
        subclasses)
        """
        return any(self._is_mixin_column(col) for col in self.columns.values())
```

taldcroft commented:
@mhvk - tl;dr: the current logic is correct and lots of stuff breaks with the swap you suggest.

Granted, it's always confusing even to me.  The `_is_mixin_column` method does this:  Determine if ``col`` meets the protocol for a mixin Table column for this table.  

For a QTable, a Quantity instance will stay as Quantity when added to the table, hence it *IS* a mixin column.  Conversely, for a Table, a Quantity instance will be converted to a plain Column, so it *IS NOT* a mixin column.

Maybe the method should be named `_will_be_a_mixin_column_when_added_to_table()`.

mhvk commented:
@taldcroft - OK, that makes sense, though indeed the name is confusing.

More to the point: I think for a `QTable`, one should not list the `Quantity` class if all columns are `Quantity`, since that is really the "expected" column. I'll make a suggestion in-line on how to change it...

mhvk commented on astropy/table/info.py:
```diff
@@ -0,0 +1,128 @@
+        # Remove 'class' info column if all table columns are the same class
+        # and they are not mixin columns.
+        uniq_types = set(type(col) for col in cols)
+        if len(uniq_types) == 1 and not tbl._is_mixin_column(cols[0]):
```
OK, the simplest would just bet:
```
if len(uniq_types) == 1 and not tbl._is_mixin_column(cols[0], quantity_is_mixin=False):
```
It is sufficiently hard to get a `Quantity` into a regular table that I think it is OK that we do not do the right thing for a regular table with only `Quantity` columns....

taldcroft replied:
Here is a logically equivalent version that is less obscure, but still relies on the fact that one *cannot* have a Quantity-mixin in a `Table` (it will always be converted to `Column`):
```
    if 'class' in info.colnames:
        # Remove 'class' info column if all table columns are the same class
        # and they are either the default class for that table or else Quantity.
        uniq_types = set(type(col) for col in cols)
        if len(uniq_types) == 1 and isinstance(cols[0], (tbl.ColumnClass, u.Quantity)):
            del info['class']
```
When I wrote this I started thinking "why is Quantity so special?".  It's just another mixin class, and has some of the same failings that other mixins have, most notably lack of masking.  The fact that there is a QTable is only to differentiate what to do with a thingy-with-units when added to a table (Quantity or Column).  

But in the end Quantity is just another mixin along with the rest, and I think it's important to always highlight this distinction by explicitly showing the class just as we would for a QTable full of Coordinates.

I'm inclined to go with:
```
        if len(uniq_types) == 1 and isinstance(cols[0], tbl.ColumnClass):
```
In the end if it's just horrible this is a soft API (just an informational output) so we can change it.

mhvk commented:
@taldcroft - OK, am convinced! I like the logic that if a table has just the type of columns it expects, we remove the `class` entry.

mhvk commented:
@taldcroft - I had a final look and it is great!  I would have merged, too, had it not been for the appveyor. It is something in `access_table.rst`, but I must admit I do not understand what is wrong...

taldcroft commented:
Oh my, frustration.  Will try once more skipping the doc tests that might be causing the problem.  Hopefully we can revert to running the doc tests later, but at least we know they did pass on linux and mac.

mhvk commented:
Weird bug.  But :fireworks: for having this!

cdeil commented:
@taldcroft – Thanks! This is great, I'll use this all the time for interactive work with tables.

If I try the Fermi catalog example again, I get a bunch or warnings:
https://gist.github.com/cdeil/97a7be42096cfe71f5c5#file-gistfile1-txt-L95
Is it important to have these warnings for this table for some reason?
Or would it be better to catch them and not show them to the end-user?

embray commented on astropy/utils/data_info.py:
```diff
@@ -0,0 +1,303 @@
+    return str(val)
+
+
+class InfoDescriptor(object):
```
I left a comment on some commit somewhere questioning the need for this class, versus just making `DataInfo` itself the descriptor.  But now I can't find that original comment anymore because GitHub.

Anyways, I see now why it was done this way--it seems to me you wanted to be able to instantiate a DataInfo only when this attribute is accessed via an instance of the class that this descriptor belongs to, and then to attach the `_parent_ref`.

Instead, if `DataInfo` (and friends) are themselves descriptors, you can still have an instance of the descriptor that lives in the class definition like:
```python
class Table(object):
    info = TableInfo()
```
this `TableInfo` is referred to as an *unbound descriptor* similarly to "unbound method".  It's like a template for later descriptors that will be *bound* to individual instances of the class.  Accessing most attributes or methods of the unbound descriptor should probably fail (for example by checking if `self._parent_ref is None`).  Though the unbound descriptor could still have some use--for example its `__repr__` can display what attributes the info object carries.

The `__get__` method can then return a *bound* version of the descriptor of the same class, the only difference being that it has its `._parent_ref` set.  Incidentally the `__get__` implemented below will break when you try to look at the descriptor through the class:

```python
In [1]: from astropy.table import Table

In [2]: Table.info
ERROR: AttributeError: 'NoneType' object has no attribute '__dict__' [astropy.utils.data_info]
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-2-915a8e227a9a> in <module>()
----> 1 Table.info

/internal/1/root/src/astropy/astropy/astropy/utils/data_info.py in __get__(self, instance, owner_cls)
    154 
    155     def __get__(self, instance, owner_cls):
--> 156         if 'info' not in instance.__dict__:
    157             instance.__dict__['info'] = self.info_cls()
    158         instance.__dict__['info']._parent = instance

AttributeError: 'NoneType' object has no attribute '__dict__'
```

This is because accessing a descriptor through the class goes through `__get__` too, but `instance` is `None` in this case.  It's typical for this case to just:
```python
if instance is None:
    return self
```
where `self` is the unbound descriptor.

Then, to make the bound descriptor you do something like:
```
info = instance.__dict__.get('info')
if info is None:
    info = self.__class__()  # Or copy or whatever
    info._parent_ref = weakref.ref(instance)
    instance.__dict__['info'] = info

return info
```

The one downside to the above approach is that the `__init__` of the info class can't rely on `_parent_ref`.  If this is undesirable you could also use the ugly but perfectly valid approach of:
```
info = self.__new__(self.__class__)
info._parent_ref = weakref.ref(instance)
info.__init__()
```
else:
    inst = instance.__dict__['info']

mhvk replied:
@embray - it would seem more elegant to make `DataInfo` itself a descriptor, following the procedure you describe. I must admit, though, I still do not fully understand it. In particular, would it also solve the large speed problems we had when `info` is present? Eg.,  in `Quantity.__array_finalize__`, can we still recognize that the descriptor is as yet uninitialized? (From the above it does seem so...). If so, to avoid loosing this again, could you repost your comment as a new issue, so it doesn't get lost again?

taldcroft replied:
> The one downside to the above approach is that the __init__ of the info class can't rely on _parent_ref.

I'm not sure what you mean exactly by "can't rely on", but definitely `_parent_ref` needs to be set every time `__get__` is called.  I had some issues along the way copying / transferring `info` and doing this on-the-fly parent referencing made things work better.  Apart from the bug with accessing `info` on the class object (I knew about this but forgot to address it), is there any problem in your approach of moving the `info._parent_ref = weakref.ref(instance)` statement outside the `if` block?

I also don't understand how the "ugly but perfectly valid" approach is different from the first.  I'm expecting another lesson in advanced Python.  :smile:

I guess it will be more elegant to be able to say `info = QuantifyInfo()` in the class definition, but we'll definitely need good code comments so mortals (and myself) of the future can understand what's happening.  I think my implementation is more easily understood by being very obvious about what is the info container and what is the descriptor.

embray replied:
What I mean is that with the bound/unbound descriptor pattern, if you want to be able to something in the class's `__init__`, that `__init__` needs to handle being called in two different contexts:

1) class definition, in which case you're initializing an unbound instance of the descriptor--this is the one that lives on the class itself and provides the descriptor protocol:
```python
class MyClass:
    info = DataInfo()  # <-- DataInfo.__init__ called here
```
2) Creating a bound instance--when `__get__` is called via the descriptor protocol, the idea is that you then create a *new* `DataInfo` instance that is bound to some instance of the parent class.  So if you want the `DataInfo.__init__` to be able to do something different for bound instances you have to check in the `__init__` for if `_parent_ref` is set or not.

An alternative is to use different subclasses for the Unbound and Bound versions.  That might be simpler to see.

See how the [`astropy.modeling.Parameter`](http://docs.astropy.org/en/stable/api/astropy.modeling.Parameter.html#astropy.modeling.Parameter) class is implemented for reference.  I think ultimately I'm trying to point out that since we already use this pattern elsewhere in Astropy it would make sense to use generally where appropriate.  Maybe I can figure out a way to make it easy to implement this pattern generically, like I did with [`OrderedDescriptor`](http://docs.astropy.org/en/latest/api/astropy.utils.misc.OrderedDescriptor.html#astropy.utils.misc.OrderedDescriptor)

embray replied:
Maybe a simple way to state this pattern is:

> Say what you mean.

So instead of the indirection of:

```python
class MyClass:
    info = InfoDescriptor(DataInfo)
```
which says "info is a thing that somehow when I access it will be a `DataInfo` object".  It forces the reader to have to think about descriptors and how that's all implemented.  Where what you really want is to be able to write:

```python
class MyClass:
    info = DataInfo()
```
which you can read as "info is a `DataInfo`".
